### PR TITLE
Refactor agent to use Observable node info

### PR DESCRIPTION
**Pull request by SpComb**
Replace the event-based `Celluloid::Notifications` `subscribe('agent:node_info')` with a new reactive `Observable` -> `Observer` interface that allows actors to restart after a crash and immediately pick up the existing state.

### `StatsWorker`

```
kontena-agent | I, [2017-03-22T09:18:14.924600 #1]  INFO -- Kontena::Workers::NodeInfoWorker: initialized
kontena-agent | I, [2017-03-22T09:18:14.948380 #1]  INFO -- Kontena::Workers::StatsWorker: initialized
kontena-agent | I, [2017-03-22T09:18:14.949128 #1]  INFO -- Kontena::Workers::StatsWorker: waiting for cadvisor
kontena-agent | I, [2017-03-22T09:18:14.971066 #1]  INFO -- Kontena::Workers::StatsWorker: cadvisor is running, starting stats loop
kontena-agent | D, [2017-03-22T09:18:14.972339 #1] DEBUG -- Workers::NodeInfoWorker: observe: #<Kontena::Workers::StatsWorker:0x005639615085e8>.node=
kontena-agent | D, [2017-03-22T09:18:16.865729 #1] DEBUG -- RpcServer: rpc notification: Kontena::Rpc::AgentApi#master_info [{"version"=>"1.2.0.dev"}]
kontena-agent | D, [2017-03-22T09:18:16.865977 #1] DEBUG -- RpcServer: rpc notification: Kontena::Rpc::AgentApi#node_info [{"id"=>"XI4K:NPOL:EQJ4:S4V7:EN3B:DHC5:KZJD:F3U2:PCAN:46EV:IO4A:63S5", "created_at"=>"2016-10-27 08:15:31 UTC", "updated_at"=>"2017-03-20 14:40:38 UTC", "name"=>"core-01", "labels"=>["test"], "overlay_ip"=>"10.81.0.1", "peer_ips"=>["192.168.66.102"], "node_number"=>1, "initial_member"=>true, "grid"=>{"id"=>"development", "name"=>"development", "initial_size"=>1, "trusted_subnets"=>[], "subnet"=>"10.81.0.0/16", "supernet"=>"10.80.0.0/12", "stats"=>{"statsd"=>{"server"=>"127.0.0.2", "port"=>8125}}}}]
kontena-agent | D, [2017-03-22T09:18:16.868890 #1] DEBUG -- Workers::NodeInfoWorker: publishing node information
kontena-agent | D, [2017-03-22T09:18:22.163605 #1] DEBUG -- Workers::NodeInfoWorker: update: #<Node:0x005639618e7d30>
kontena-agent | D, [2017-03-22T09:18:22.164576 #1] DEBUG -- Workers::NodeInfoWorker: notify: #<Kontena::Workers::StatsWorker:0x005639615085e8>.node= #<Node:0x005639618e7d30>
kontena-agent | D, [2017-03-22T09:18:22.164889 #1] DEBUG -- Workers::StatsWorker: configure statds: {"server"=>"127.0.0.2", "port"=>8125}
kontena-agent | I, [2017-03-22T09:18:22.164971 #1]  INFO -- Kontena::Workers::StatsWorker: exporting stats via statsd to udp://127.0.0.2:8125
```

The `StatsWorker` now recovers quickly from a crash, without having to wait for a `node.info` notification:

```
kontena-agent | D, [2017-03-22T09:19:15.076450 #1] DEBUG -- Workers::StatsWorker: starting collection
kontena-agent | E, [2017-03-22T09:19:15.206125 #1] ERROR -- : Actor crashed!
kontena-agent | RuntimeError: crash
kontena-agent | 	/app/lib/kontena/workers/stats_worker.rb:46:in `block in start'
kontena-agent | 	/app/lib/kontena/workers/stats_worker.rb:41:in `loop'
kontena-agent | 	/app/lib/kontena/workers/stats_worker.rb:41:in `start'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
kontena-agent | I, [2017-03-22T09:19:15.221604 #1]  INFO -- Kontena::Workers::StatsWorker: initialized
kontena-agent | I, [2017-03-22T09:19:15.221883 #1]  INFO -- Kontena::Workers::StatsWorker: waiting for cadvisor
kontena-agent | I, [2017-03-22T09:19:15.258843 #1]  INFO -- Kontena::Workers::StatsWorker: cadvisor is running, starting stats loop
kontena-agent | D, [2017-03-22T09:19:20.002071 #1] DEBUG -- Workers::NodeInfoWorker: observe: #<Kontena::Workers::StatsWorker:0x005639623c3f60>.node=
kontena-agent | D, [2017-03-22T09:19:20.002294 #1] DEBUG -- Workers::StatsWorker: configure statds: {"server"=>"127.0.0.2", "port"=>8125}
kontena-agent | I, [2017-03-22T09:19:20.002349 #1]  INFO -- Kontena::Workers::StatsWorker: exporting stats via statsd to udp://127.0.0.2:8125
```

### `FluentdWorker`
```
kontena-agent | D, [2017-03-22T10:03:58.274701 #1] DEBUG -- Workers::NodeInfoWorker: update: #<Node:0x0055b7fc7f4838>
kontena-agent | D, [2017-03-22T10:03:58.277032 #1] DEBUG -- Workers::NodeInfoWorker: notify: #<Kontena::Workers::FluentdWorker:0x0055b7fd37c818>.node= #<Node:0x0055b7fc7f4838>
kontena-agent | I, [2017-03-22T10:03:58.277313 #1]  INFO -- Kontena::Workers::FluentdWorker: starting fluentd log streaming to 127.0.0.2
```

`kontena grid update --log-forwarder none development`
```
kontena-agent | D, [2017-03-22T10:04:07.773692 #1] DEBUG -- RpcServer: rpc notification: Kontena::Rpc::AgentApi#node_info [{"id"=>"XI4K:NPOL:EQJ4:S4V7:EN3B:DHC5:KZJD:F3U2:PCAN:46EV:IO4A:63S5", ..., "logs"=>nil}}]
kontena-agent | D, [2017-03-22T10:04:07.775639 #1] DEBUG -- Workers::NodeInfoWorker: update: #<Node:0x0055b7fe67e1a0>
kontena-agent | D, [2017-03-22T10:04:07.784650 #1] DEBUG -- Workers::NodeInfoWorker: notify: #<Kontena::Workers::FluentdWorker:0x0055b7fd37c818>.node= #<Node:0x0055b7fe67e1a0>
kontena-agent | I, [2017-03-22T10:04:07.785990 #1]  INFO -- Kontena::Workers::FluentdWorker: stopping fluentd log streaming
```

### CODE REVIEW

**Comment by jakolehm** [comment]
Design review conclusion:

```
observe(Actor[:node_info_worker], Actor[:etcd_launcher]) do |node, etcd_status|
  ensure_config(node)
end
```

**Comment by jakolehm** [comment]
@SpComb any progress here?

**Comment by SpComb** [comment]
Reimplemented using the raw `Celluloid::Mailbox` with a `Observable::Message` containing a separate `state` handle for each `observe` task.

One more thing to decide on: is `observe` itself sync or async? The previous model of `observe` being sync and calling the block directly is broken: when used in an Actor's `initialize` method, this would yield to the block directly from the `initialize` method, which could block.

Does each Actor start a new `async` task to run call `observe`, or should `observe` itself make a new async task to yield to the block from?

**Comment by jnummelin** [comment]
From pure API point of view I think it would be cleaner if `async` is handled within `observe`. This makes the user side of observers a bit cleaner IMO.

**Comment by SpComb** [comment]
Design changes implemented.

File: agent/lib/kontena/helpers/observable.rb [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        value = observe.set(observable, observable.add_observer(Celluloid.current_actor, observe))
+
+        debug "observe #{observable} -> #{value}"
+
+        # crash if observed Actor crashes, otherwise we get stuck without updates
+        self.link observable
+      end
+
+      # async update message loop
+      Thread.current[:celluloid_actor].task :observe do
```
Comment (SpComb) [inline agent/lib/kontena/helpers/observable.rb]: This is based on how Celluloid implements e.g. timers, but maybe it's too much into the internals... Use `self.async._obseve_task(observe)` instead?

**Comment by SpComb** [comment]
Replaced the mailbox stuff with `Observe` state stored in the `Observable#observers`, used by the `Observer` to pick the right `observe` block to run when updating.

File: agent/lib/kontena/helpers/observable.rb [hunk 2]
```diff
@@ -0,0 +1,167 @@
+module Kontena
+  # An Actor that has some value
+  # The value does not yet exist when initialized, it is nil
+  # Once the value is first updated, then other Actors will be able to observe it
+  # When the value later updated, other Actors will also observe those changes
+  module Observable
```
Comment (jakolehm) [inline agent/lib/kontena/helpers/observable.rb]: Module namespace does not match filesystem.. I think we could put this directly under `lib/kontena/`.

File: agent/lib/kontena/helpers/observable.rb [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+          observer.async.update_observed(observe, Celluloid.current_actor, @value)
+        rescue Celluloid::DeadActorError => error
+          observers.delete(observe)
+        end
+      end
+    end
+  end
+
+  # An Actor that observes the value of other Obervable Actors
+  module Observer
```
Comment (jakolehm) [inline agent/lib/kontena/helpers/observable.rb]: We should probably split this into own file?

File: agent/lib/kontena/workers/stats_worker.rb [hunk 4]
```diff
@@ -151,7 +156,7 @@ def send_statsd_metrics(name, event)
       if labels && labels[:'io.kontena.service.name']
         key_base = "services.#{name}"
       else
-        key_base = "#{node_name}.containers.#{name}"
+        key_base = "#{@node_name}.containers.#{name}"
```
Comment (jakolehm) [inline agent/lib/kontena/workers/stats_worker.rb]: Why not `node.name`?

Comment (SpComb) [inline agent/lib/kontena/workers/stats_worker.rb]: This is in a different method, called much later after `configure_statsd`.

File: agent/lib/kontena/observable.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
+module Kontena
+  # An Actor that has some value
+  # The value does not yet exist when initialized, it is nil
+  # Once the value is first updated, then other Actors will be able to observe it
+  # When the value later updated, other Actors will also observe those changes
+  module Observable
+    include Kontena::Logging
+
+    def value
```
Comment (jakolehm) [inline agent/lib/kontena/observable.rb]: Probably too common name to have in a included module? Maybe `observable_value`?

File: agent/lib/kontena/observable.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      @observers ||= {}
+    end
+
+    # The Observable has a value.
+    #
+    # This will notify any Observers, causing them to yield if ready.
+    #
+    # @param value [Object]
+    # @raise [ArgumentError] Update with nil value
+    def update(value)
```
Comment (jakolehm) [inline agent/lib/kontena/observable.rb]: Also too common?

File: agent/lib/kontena/observable.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      debug "update: #{value}"
+
+      @value = value
+
+      notify_observers
+    end
+
+    # The Observable no longer has a value
+    # This will notify any Observers, causing them to block yields until we update again
+    def reset
```
Comment (jakolehm) [inline agent/lib/kontena/observable.rb]: Also too common?

File: agent/lib/kontena/workers/service_pod_manager.rb [hunk 8]
```diff
@@ -16,26 +17,25 @@ class ServicePodManager
 
     def initialize(autostart = true)
       @workers = {}
-      subscribe('agent:node_info', :on_node_info)
-      subscribe('service_pod:update', :on_update_notify)
-      async.start if autostart
+
+      if autostart
+        observe(Actor[:node_info_worker]) do |node|
```
Comment (jakolehm) [inline agent/lib/kontena/workers/service_pod_manager.rb]: Should this observe be inside `#start` .. how can actor start observing if `autostart=false` and start is called manually?

File: agent/lib/kontena/workers/stats_worker.rb [hunk 9]
```diff
@@ -3,25 +3,30 @@ class StatsWorker
...
+    attr_reader :statsd
 
     # @param [Boolean] autostart
     def initialize(autostart = true)
       @statsd = nil
-      @node_name = nil
       info 'initialized'
-      subscribe('agent:node_info', :on_node_info)
       async.start if autostart
+
```
Comment (jakolehm) [inline agent/lib/kontena/workers/stats_worker.rb]: Double `if autostart`

File: agent/lib/kontena/workers/stats_worker.rb (hunk 4 above)
Comment (jakolehm) [inline agent/lib/kontena/workers/stats_worker.rb]: Oh, right... but it smells a bit to have `@node_name` instance variable.. why not store whole `@node`?

File: agent/lib/kontena/observable.rb (hunk 5 above)
Comment (SpComb) [inline agent/lib/kontena/observable.rb]: Yes, this isn't really even exposed.

File: agent/lib/kontena/observable.rb (hunk 6 above)
Comment (SpComb) [inline agent/lib/kontena/observable.rb]: `update_observable(...)` and `reset_observable`? It's a little verbose, but not too much.

File: agent/lib/kontena/workers/service_pod_manager.rb [hunk 10]
```diff
@@ -15,27 +16,26 @@ class ServicePodManager
...
+
       async.start if autostart
     end
 
     def start
+      observe(Actor[:node_info_worker]) do |node|
+        @node = node
+      end
+
       wait_until!("have node info", interval: 0.1) { self.node }
```
Comment (SpComb) [inline agent/lib/kontena/workers/service_pod_manager.rb]: This is still kinda clunky... but the reason is that this should eventually look something like

```ruby
observe(Actor[:node_info_worker], Actor[:service_pods_worker], Actor[:docker_containers_worker]) do |node, service_pods, docker_containers|
   ....
end
```

**Pull request merged**